### PR TITLE
[Misc] NFC: Fix -Wdefaulted-function-deleted warnings

### PR DESCRIPTION
**Pull request by davezarzycki**
Top-of-tree clang enables `-Wdefaulted-function-deleted`, therefore the following changes are needed for warning free builds.

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

File: include/swift/AST/RawComment.h [hunk 1]
```diff
@@ -38,7 +38,9 @@ struct SingleRawComment {
   SingleRawComment(StringRef RawText, unsigned StartColumn);
 
   SingleRawComment(const SingleRawComment &) = default;
-  SingleRawComment &operator=(const SingleRawComment &) = default;
+
+  // 'Range' has no copy assignment operator; therefore:
```
Comment (jrose-apple) [inline include/swift/AST/RawComment.h]: That seems like a bug, actually. CharSourceRange should be a trivial type. Or is this about the `const`?

File: lib/SILGen/LValue.h [hunk 2]
```diff
@@ -524,7 +524,9 @@ struct LLVM_LIBRARY_VISIBILITY ExclusiveBorrowFormalAccess : FormalAccess {
   ExclusiveBorrowFormalAccess &
   operator=(ExclusiveBorrowFormalAccess &&) = default;
 
-  ExclusiveBorrowFormalAccess() = default;
+  // 'swift::Lowering::FormalAccess' has no default constructor; therefore:
+  ExclusiveBorrowFormalAccess() = delete;
```
Comment (jrose-apple) [inline lib/SILGen/LValue.h]: I think we'd probably just leave this out.

**Review by jrose-apple — COMMENTED** [review]
I feel like many of these are indicative of other bugs, and should not necessarily be silenced in this way. But maybe that just means tweaking the comments to include "FIXME" or something.

I don't think we should ever have a case where something has a move constructor but not a move assignment operator except if there's a very good reason. If it's just `const` members, we should un-constify them unless it's semantically important that all the const members are initialized together.

File: include/swift/AST/RawComment.h (hunk 1 above)
Comment (davezarzycki) [reply include/swift/AST/RawComment.h]: 'const' does seem to be the problem.

**Comment by davezarzycki** [comment]
Hi @jrose-apple – I was worried about this devolving into something like "const correctness" whack-a-mole, but it turns out to be not that bad. Please consider the updated patch.

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@davezarzycki thanks for fixing this, I had gotten half way through the same set of changes and then got worried how deep the rabbit hole was.

**Review by compnerd — APPROVED** [review]
I had tried to use nullability attributes for the ProjectionTree to provide the same guarantees, but that got out of hand pretty quick.

File: include/swift/Basic/SourceLoc.h [hunk 3]
```diff
@@ -130,7 +130,9 @@ class CharSourceRange {
 
 public:
   /// \brief Constructs an invalid range.
-  CharSourceRange() {}
+  CharSourceRange() = default;
+
+  CharSourceRange &operator=(const CharSourceRange &) = default;
```
Comment (jrose-apple) [inline include/swift/Basic/SourceLoc.h]: [Rule of Zero](https://en.cppreference.com/w/cpp/language/rule_of_three) says that we should not declare this explicitly. If you think it's important, we should follow the Rule of Five instead.

File: include/swift/Basic/TreeScopedHashTable.h [hunk 4]
```diff
@@ -37,6 +37,11 @@ template <typename K, typename V> class TreeScopedHashTableVal {
   TreeScopedHashTableVal(const K &Key, const V &Val) : Key(Key), Val(Val) {}
 
 public:
+  TreeScopedHashTableVal(const TreeScopedHashTableVal &) = delete;
+  TreeScopedHashTableVal(TreeScopedHashTableVal &&) = delete;
+  TreeScopedHashTableVal &operator=(const TreeScopedHashTableVal &) = delete;
+  TreeScopedHashTableVal &operator=(TreeScopedHashTableVal &&) = delete;
```
Comment (jrose-apple) [inline include/swift/Basic/TreeScopedHashTable.h]: Rule of Five says it's worth mentioning the destructor here too.

File: lib/SILGen/LValue.h [hunk 5]
```diff
@@ -524,7 +524,6 @@ struct LLVM_LIBRARY_VISIBILITY ExclusiveBorrowFormalAccess : FormalAccess {
   ExclusiveBorrowFormalAccess &
   operator=(ExclusiveBorrowFormalAccess &&) = default;
 
-  ExclusiveBorrowFormalAccess() = default;
```
Comment (jrose-apple) [inline lib/SILGen/LValue.h]: Why this change?

Comment (davezarzycki) [reply lib/SILGen/LValue.h]: The superclass doesn't have a default constructor, so the subclass can't either.

File: include/swift/Basic/SourceLoc.h (hunk 3 above)
Comment (davezarzycki) [reply include/swift/Basic/SourceLoc.h]: `CharSourceRange() {}` is effectively `CharSourceRange() = default`, right? If so, then yes, I can flush out the rule-of-five.
